### DISCUSSION TITLE
Terratech Schematic Calculation

### DISCUSSION THREAD

**Discussion by jojoblackFr** [AsherGlick/ResourceCalculator#137] [Ideas]
I wonder, Is it possible to create a plugins for a specific calculator ? 
I want to create a plugins that take a schematic and output the necessary data (block required for the schematic) for the calculator to do it's job.
Maybe even add a way to send the inventory of the player from the game into the calculator from it so you don't over craft

If it's possible how can I ?

**Comment by AsherGlick** [comment]
We have one plugin currently for importing litematica minecraft data into the resource calculator. It is the only "plugin" we have now, they just get copied over directly to the output without processing at the moment.. https://github.com/AsherGlick/ResourceCalculator/tree/master/resource_lists/minecraft/plugins/litematica . This solution is a little hacky and plugins will need to be supported by the community if it ever breaks, I dont plant to update them directly myself, but lots of folks use this litematica one and it has been running well for a while.

There was also some talk of adding in a way to import csv files or some other data file into the calculator, https://github.com/AsherGlick/ResourceCalculator/issues/96 but that will require some design work to figure out how it should look and feel before it can be implemented.

As far as exporting the inventory from the game, that is going to be game dependent, there will probably need to be a unique mod or script made for the game to support that.

**Reply by jojoblackFr** [reply]
Basically to create a plugin it's just another page that redirect on the main calculator with the data. Indeed "hacky".

The CSV file is a great idea but not every game decide to save in CSV. I would suggest that it take any file as an input and it will execute a plugins made with the calculator in the resource folder like `resource_list/xxx/plugins.js` with only one function, and if there is no `plugins.js` files then it will just disable the `select files` options to not break the already existing calculator.

As for the inventory it's just some minor details, my main goal is the schematic import.

Welp, time to take a look into it

**Reply by AsherGlick** [reply]
We used to have that feature, allowing for importing of minecraft nbt data. However it became too much work to maintain that system, we got too many complaints about it not working, and it was eventually dropped. The current plugin system is a compromise to still allow some custom behavior, but also not support it or advertise it to users who are unaware of its existence.

**Reply by jojoblackFr** [reply]
Basically becoming a mouth to ear kind of information.
Thx for the answer
